xapian-applications/omega/omindexcheck.cc

```diff
@@ -0,0 +1,142 @@
+    tests.insert({"fb2/lang2.fb2", {"Adavid", "Atardon", "ZAtardon",
+		  "Zúhořů", "Zhovercraft", "Zlanguag", "Zmój", "Zof",
+		  "Zpełen", "Zpoduszkowiec", "Zvznášedlo", "Zwęgorzi"}});
+    tests.insert({"lrf/hello.lrf", {"ZFhello", "Zhello", "Zworld"}});
+    tests.insert({"pdb/PalmDOC-hello.pdb", {"Shello", "Sworld", "ZFpalmdoc",
+		  "Zworld"}});
+    tests.insert({"pdb/PeanutPress-hello.pdb", {"Fpeanutpress", "ZFhello",
+		  "Zhello"}});
+    tests.insert({"pdb/test.pdb", {"Sdemodemo", "ZFtest", "Zwherearew"}});
+#endif
+#if defined HAVE_LIBETONYEK
+    tests.insert({"keynotes/test-keynote.key", {"ZFkeynot", "Zbold", "Znow",
+		  "Zsubtitl"}});
+    tests.insert({"pages/test-pages.pages", {"ZFpage", "Zfull", "Zhovercraft",
+		  "Zwęgorzi"}});
+#endif
+#if defined HAVE_TESSERACT
+    tests.insert({"img/Test1.gif", {"Znoisyimag", "Zocr", "Ztesseract"}});
+    tests.insert({"img/Test2.pgm", {"ZFtest2", "Znoisyimag", "Ztesseract"}});
+    tests.insert({"img/Test3.ppm", {"ZFtest3", "Zocr", "Ztest"}});
+    tests.insert({"img/Test4.tiff", {"Znoisyimag", "Ztesseract", "Zto"}});
+    tests.insert({"img/Test5.webp", {"Znoisyimag", "Ztesseract", "Ztest"}});
+    tests.insert({"img/poster-2.jpg", {"ZFposter", "Zby", "Zproperti",
+		  "Zsurveil", "Zvideo"}});
+    tests.insert({"img/poster.jpg", {"Zbicycl", "Zride", "Zroller",
+		  "Zskateboard"}});
+    tests.insert({"img/scan-page.png", {"Zannual", "Zfed", "Zreturn",
+		  "Zwhile"}});
+#endif
+}
+
+static bool
+compare_test(testcase& test, const Xapian::Document& doc, const string& file)
+{
+    sort(test.begin(), test.end());
+    Xapian::TermIterator term_iterator = doc.termlist_begin();
+    for (auto& t : test) {
+	term_iterator.skip_to(t);
+	if (term_iterator == doc.termlist_end() || *term_iterator != t) {
+	    cerr << "Error in " << file << ": Term " << t <<
+		 " does not belong to this file" << endl;
+	    return false;
+	}
+    }
+    return true;
+}
+
+int
+main(int argc, char** argv)
+{
+    Xapian::Database db;
+    bool succeed = true;
+    if (argc <= 2)
+	return 1;
+    db.add_database(Xapian::Database(argv[2]));
+    string url, current_dir(argv[1]);
+
+    index_test();
+    for (auto t = db.allterms_begin("U"); t != db.allterms_end("U"); ++t) {
+	const string& term = *t;
+	Xapian::PostingIterator p = db.postlist_begin(term);
+	if (p == db.postlist_end(term)) {
+	    continue;
+	}
+	Xapian::docid did = *p;
+	Xapian::Document doc = db.get_document(did);
+	auto data = doc.get_data();
+	size_t start;
+	if (startswith(data, "url=")) {
+	    start = CONST_STRLEN("url=");
+	} else {
+	    start = data.find("\nurl=");
+	    if (start == string::npos)
+		continue;
+	    start += CONST_STRLEN("\nurl=");
+	}
+	url.assign(data, start, data.find('\n', start) - start);
+	start = url.find(current_dir) + current_dir.length();
+	url = url.substr(start, url.length());
+	auto it = tests.find(url);
+	if (it != tests.end())
+	    succeed &= compare_test(it->second, doc, url);
```

Comment:
If there's no entry in `tests` found corresponding to `url` then we quietly ignore that and keep going - shouldn't that situation cause the tests to fail?

Reply:
Not sure if we should fails here.
I think that the main(or first) idea of this pr was to test the code of some handlers and run them during the CI build. That is why I used conditional compilation flags to add(or not) the cases for each file. This way, I just compared the test cases if I was sure that the workers were available in the system. 

As there may be more than one way to index the same file in some systems, it is possible that some workers were not available and omindex indexed some files anyway. That means that some term could be missing and `omindexcheck` would fail.

Maybe we can think in something more "robust" and remove the conditional flags. That way we check that if a file was indexed by omindex, it will have at least the terms of `omindexcheck`. However, we will need to be really careful when choosing the terms as they should be always present in the documents.

Reply:
Oh I see - your thinking that (for example) libarchive isn't available and then we fall back to indexing the file a different way, which might not result in exactly the same terms.

We could probably just disable indexing of the extensions in question in `#else` of the conditions where we add the testcases.

My concerns here are really that it's currently possible to half-add a testcase and not realise that your new test isn't actually being checked, and that a bug which results in a document being misindexed might not be flagged.  I think ideally we would also check that all the testcases which got enabled were checked exactly once.

Or perhaps we should just let the documents be indexed anyway and this serves as a check of whatever method is used.  As you say, we'd might need to select the terms to check with more care.

I'm OK to leave this as-is for now in the interests of getting this merged and usable for building on.

Reply:
Yes, I agree. Maybe some improvements could be done, but I think it is okay as a first check.